gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java

```diff
@@ -0,0 +1,351 @@
+    lock.lock();
+
+    try {
+      return this.client.tableExists(dbName, tableName);
+    } catch (TException e) {
+      log.error(String.format("Unable to check existence of table %s in db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public boolean existsPartition(String dbName, String tableName, List<String> partitionValues) {
+    Lock lock = locks.get(dbName + tableName);
+    lock.lock();
+
+    try {
+      client.getPartition(dbName, tableName, partitionValues);
+      return true;
+    } catch (NoSuchObjectException e) {
+      return false;
+    } catch (TException e) {
+      log.error(String.format("Unable to check existence of partition %s in table %s in db %s", partitionValues,
+          tableName, dbName));
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public void dropTable(String dbName, String tableName) {
+    Lock lock = locks.get(dbName + tableName);
+    lock.lock();
+
+    try {
+      if (this.client.tableExists(dbName, tableName)) {
+        this.client.dropTable(dbName, tableName);
+        log.info("Dropped table " + tableName + " in db " + dbName);
+      }
+    } catch (TException e) {
+      log.error(String.format("Unable to deregister table %s in db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public void dropPartition(String dbName, String tableName, List<String> partitionValues) {
+    Lock lock = locks.get(dbName + tableName);
+    lock.lock();
+
+    try {
+      client.dropPartition(dbName, tableName, partitionValues, false);
+      log.info("Dropped partition " + partitionValues + " in table " + tableName + " in db " + dbName);
+    } catch (NoSuchObjectException e) {
+      // Partition does not exist. Nothing to do
+    } catch (TException e) {
+      log.error(String.format("Unable to check existence of Hive partition %s in table %s in db %s", partitionValues,
+          tableName, dbName));
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private void addPartition(Table table, HivePartition partition, Path partitionLocation) throws TException {
+    Preconditions.checkArgument(table.getPartitionKeysSize() == partition.getValues().size());
+
+    Partition p = new Partition();
+    p.setDbName(table.getDbName());
+    p.setTableName(table.getTableName());
+    p.setValues(partition.getValues());
+    p.setSd(table.getSd());
+    p.getSd().setLocation(partitionLocation.toString());
+
+    Lock lock = locks.get(table.getDbName() + table.getTableName());
+    lock.lock();
+
+    try {
+      this.client.add_partition(p);
+      log.info(String.format("Added partition %s to table %s with location %s", partition, table.getTableName(),
+          p.getSd().getLocation()));
+    } catch (AlreadyExistsException e) {
+      // Partition already exists. Nothing to do.
```

Comment:
I think it's worth emitting a log message about this. Also, what if the partition that already exists has different values for the StorageDescriptor, location, etc?